### PR TITLE
ek/reprint

### LINKED ISSUE #73: Scripts: collect and reprint warnings at build finish
Use [following](http://stackoverflow.com/questions/31348508/save-and-reprint-warnings-for-successfully-compiled-files-on-subsequent-builds) technique to reprint warnings.

* do this only for Debug build;
* make it work without any external scripts (like `build.sh` in answer);
* review code in `capture_warnings.sh.in` and fix wrong points:
    - no need to generate `remove_obj_with_warnings.sh`;
    - no need to output per-object `.warnings` files, there should be one single file in top build dir: `compile.warnings`;
    - `make clean` removes `compile.warnings`.

### PR DESCRIPTION
**Pull request by kevgs**
Closes #73

### CODE REVIEW

**Comment by midenok** [comment]
Add **closes** clause.

File: BUILD/capture_warnings.sh [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+TMP_FILE=/tmp/compile.warning
```
Comment (midenok) [inline BUILD/capture_warnings.sh]: - Wrong usage of fixed file name in **/tmp**. Use `mktemp`. Use **/dev/shm** to utilize tmpfs;
- `TMP_FILE` is not needed at all (see below);
- Use lower case for non-exported variables.

File: BUILD/capture_warnings.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+WARN_PATH=$1
+shift
+WARN_FILE="$WARN_PATH/compile.warnings"
+
+set -o pipefail
+rm -f $TMP_FILE
+$@ 2> $TMP_FILE
```
Comment (midenok) [inline BUILD/capture_warnings.sh]: 1. Use this technique to catch `STDERR` to variable:
   
   ```
   exec 3>&1
   x=$((echo err >&2; echo out) 2>&1 1>&3)
   ```
2. Wrong expansion of `$@`. Use `"$@" instead`.

File: CMakeLists.txt [hunk 3]
```diff
@@ -39,6 +39,11 @@ MESSAGE(STATUS "Running cmake version ${CMAKE_VERSION}")
 
 SET(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${CMAKE_SOURCE_DIR}/cmake)
 
+SET_DIRECTORY_PROPERTIES(PROPERTIES
+  RULE_LAUNCH_COMPILE "bash ${CMAKE_SOURCE_DIR}/BUILD/capture_warnings.sh ${CMAKE_BINARY_DIR}")
+SET_DIRECTORY_PROPERTIES(PROPERTY ADDITIONAL_MAKE_CLEAN_FILES
+  "${CMAKE_BINARY_DIR}/compile.warnings")
```
Comment (midenok) [inline CMakeLists.txt]: Where is deletion at compilation start?

File: sql/CMakeLists.txt [hunk 4]
```diff
@@ -427,3 +427,9 @@ INSTALL(DIRECTORY . DESTINATION ${INSTALL_INCLUDEDIR}/private COMPONENT Developm
   FILES_MATCHING PATTERN "*.h"
   PATTERN share EXCLUDE
   PATTERN CMakeFiles EXCLUDE)
+
+ADD_CUSTOM_COMMAND(TARGET
+  mysqld
+  POST_BUILD
+  COMMAND if [ -f "${CMAKE_BINARY_DIR}/compile.warnings" ]; then cat "${CMAKE_BINARY_DIR}/compile.warnings" \; fi
+  DEPENDS mysqld)
```
Comment (midenok) [inline sql/CMakeLists.txt]: - Modification of several `CMakeLists.txt` is highly unwanted.
- p.1 of task is not done.

**Review by midenok — CHANGES_REQUESTED** [review]

**Comment by kevgs** [comment]
I failed with 'exec 3>&1' trick and used something else.

File: BUILD/capture_warnings.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+warn_path=$1
+shift
+warn_file="$warn_path/compile.warnings"
+
+set -o pipefail
+x=$(("$@") 2>&1 | tee /dev/stderr)
```
Comment (midenok) [inline BUILD/capture_warnings.sh]: Does this work correctly?

Comment (midenok) [reply BUILD/capture_warnings.sh]: Use:

``` bash
exec 3>&1
cmderr=$("$@" 2>&1 1>&3)
if [[ -n "$cmderr" ]]
    echo -n "$cmderr" >&2
...
```

Rename `x` to `cmderr`.

File: BUILD/capture_warnings.sh [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+error=${pipestatus}
+
+if ! [[ -z "$x" ]]; then
+    echo "$x" >> $warn_file
```
Comment (midenok) [inline BUILD/capture_warnings.sh]: echo -n ?

File: BUILD/capture_warnings.sh [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+warn_path=$1
+shift
+warn_file="$warn_path/compile.warnings"
+
+set -o pipefail
+x=$(("$@") 2>&1 | tee /dev/stderr)
+error=${pipestatus}
```
Comment (midenok) [inline BUILD/capture_warnings.sh]: PIPESTATUS is array variable. Use `$?` instead.

File: CMakeLists.txt [hunk 8]
```diff
@@ -480,3 +490,10 @@ IF(NON_DISTRIBUTABLE_WARNING)
   MESSAGE(WARNING "
 You have linked MariaDB with GPLv3 libraries!  You may not distribute the resulting binary. If you do, you will put yourself into a legal problem with Free Software Foundation.")
 ENDIF()
+
+IF (CMAKE_BUILD_TYPE MATCHES "Debug")
```
Comment (midenok) [inline CMakeLists.txt:484]: Can it be under single IF condition?

File: CMakeLists.txt [hunk 9]
```diff
@@ -480,3 +490,10 @@ IF(NON_DISTRIBUTABLE_WARNING)
...
+  ADD_CUSTOM_TARGET(REPRINT ALL
+    COMMAND if [ -f "compile.warnings" ]; then cat "compile.warnings" \; fi
+    DEPENDS mysql udf_example explain_filename-t
```
Comment (midenok) [inline CMakeLists.txt]: DEPENDS REMOVE_WARNINGS_FILE

**Pull request merged**